include/picongpu/plugins/openPMD/WriteSpecies.hpp

```diff
@@ -334,33 +398,106 @@ namespace openPMD
             Frame< OperatorCreateVectorBox, NewParticleDescription >;
 
         void
-        setParticleAttributes( ::openPMD::Iteration & iteration )
+        setParticleAttributes( ::openPMD::ParticleSpecies & record )
         {
             const float_64 particleShape(
                 GetShape< ThisSpecies >::type::assignmentFunctionOrder );
-            iteration.setAttribute( "particleShape", particleShape );
+            record.setAttribute( "particleShape", particleShape );
 
             traits::GetSpeciesFlagName< ThisSpecies, current<> >
                 currentDepositionName;
             const std::string currentDeposition( currentDepositionName() );
-            iteration.setAttribute(
+            record.setAttribute(
                 "currentDeposition", currentDeposition.c_str() );
 
             traits::GetSpeciesFlagName< ThisSpecies, particlePusher<> >
                 particlePushName;
             const std::string particlePush( particlePushName() );
-            iteration.setAttribute( "particlePush", particlePush.c_str() );
+            record.setAttribute( "particlePush", particlePush.c_str() );
 
             traits::GetSpeciesFlagName< ThisSpecies, interpolation<> >
                 particleInterpolationName;
             const std::string particleInterpolation(
                 particleInterpolationName() );
-            iteration.setAttribute(
+            record.setAttribute(
                 "particleInterpolation", particleInterpolation.c_str() );
 
             const std::string particleSmoothing( "none" );
-            iteration.setAttribute(
+            record.setAttribute(
                 "particleSmoothing", particleSmoothing.c_str() );
+
+	    // we do this once for a unit dimension output function, do not touch this code
+	    std::vector< float_64 > zeroUnitDimension( 7, 0.0 );
+	    static constexpr ::openPMD::UnitDimension
+	      openPMDUnitDimensions[ 7 ] = {
+	      ::openPMD::UnitDimension::L,
+	      ::openPMD::UnitDimension::M,
+	      ::openPMD::UnitDimension::T,
+	      ::openPMD::UnitDimension::I,
+	      ::openPMD::UnitDimension::theta,
+	      ::openPMD::UnitDimension::N,
+	      ::openPMD::UnitDimension::J };
+	    std::map<::openPMD::UnitDimension, double > zeroUnitMap;
+	    for( unsigned i = 0; i < 7; ++i )
+	      {
+		zeroUnitMap[ openPMDUnitDimensions[ i ] ] = zeroUnitDimension[ i ];
+	      }
```

Comment:
I think instead of this hacky map manipulation it would be better to have a function that takes the common PIConGPU output of `...::getUnit()`, which is `std::vector<double>` of size 7, and converts it to such `std::map<::openPMD::UnitDimension, double >`.

Reply:
put the helper into `plugins/output` or `plugins/misc`